File: master/server_main/server_game_module/src/Game/Entity/PlayerSkill.cs
namespace GamePlay;

public record PlayerSkill(
    int id,
    int level,
    int exp
)
{
    public bool CanUpgrade(PlayerDataManager Ctx)
    {
        GameAssert.Expect(!IsMaxLevel(Ctx), 500);
        var tbl = Ctx.Table.PlayerSkillTblMap[id];
        var quality = tbl.Quality;
        var cost = Ctx.Table.PlayerSkillLevelTblList.FirstOrDefault(t => t.Level == level);
        if (cost != null)
            return exp >= cost.Require[quality - 1];
        else return false;
    }
    public bool IsMaxLevel(PlayerDataManager Ctx)
    {
        return level > Ctx.Table.PlayerSkillLevelTblList.Length;
    }
    public int UpgradeCost(PlayerDataManager Ctx)
    {
        var quality = Ctx.Table.PlayerSkillTblMap[id].Quality;
        var tbl = Ctx.Table.PlayerSkillLevelTblList.FirstOrDefault(t => t.Level == level);
        if (tbl != null)
            return tbl.Require[quality - 1];
        else return 0;
    }
};

File: master/client/assets/script/game/entity/PlayerSkill.ts
export class PlayerSkill {
    /** id */
    id: number;
    /**等级 */
    level: number;
    /**库存 */
    exp: number;

    getTbl() {
        return GTable.getById("PlayerSkillTbl", this.id);
    }
    /**是否最大等级 */
    isMaxLevel(): boolean {
        return this.level > GTable.getList("PlayerSkillLevelTbl").length;
    }
    /**下一等级消耗 */
    nextLevelCost(): number {
        if (this.level > 0) {
            const quality = this.getTbl().quality;
            return GTable.getList("PlayerSkillLevelTbl").find((t) => t.level == this.level).require[quality - 1];
        } else return 1;
    }
    /**可否升级 */
    canUpgrade(): boolean {
        return !this.isMaxLevel() && this.exp >= this.nextLevelCost();
    }
}
